Cs status output updates

fischer-ncar opened the pull request:
Refactor code is test_status.py to remove duplicate code in phase_statuses_dump.
Add BFAIL the TestStatus comment when baselines are missing.
Set Overall status to BFAIL when test runs, but baselines are missing.

Test suite: Tested with cesm2_0_beta01 testresults
Test baseline: 
Test namelist changes: 
Test status:

Fixes #563 #573 

User interface changes?: 

Code review:

jgfouca commented on utils/python/CIME/test_status.py:
```diff
@@ -121,24 +123,22 @@ def get_status(self, phase):
-                        fd.write("%s %s %s\n" % (status, self._test_name, phase))
-                    else:
-                        fd.write("%s %s %s %s\n" % (status, self._test_name, phase, comments))
+        with open(self._filename, "w") as fd:
```
Good cleanup

jgfouca commented on utils/python/CIME/hist_utils.py:
```diff
@@ -277,7 +277,7 @@ def compare_baseline(case, baseline_dir=None):
     for bdir in dirs_to_check:
         if not os.path.isdir(bdir):
-            return False, "ERROR baseline directory '%s' does not exist" % bdir
+            return False, "ERROR BFAIL baseline directory '%s' does not exist" % bdir
```
I'm fine with this.

billsacks replied:
I think this should use the constant defined in test_status.py (currently named TEST_BFAIL_STATUS) rather than hard-coding the string 'BFAIL'.

billsacks replied:
Or, according to my comment below, this would use the TEST_NO_BASELINES_COMMENT

jgfouca commented on utils/python/CIME/test_status.py:
```diff
@@ -241,7 +242,10 @@ def get_overall_test_status(self, wait_for_run=False, check_throughput=False, ch
 
                 elif (rv in [NAMELIST_FAIL_STATUS, TEST_PASS_STATUS] and phase.startswith(COMPARE_PHASE)):
-                    rv = TEST_DIFF_STATUS
+                    if (phase.startswith(COMPARE_BASELINE_PHASE) and TEST_BFAIL_STATUS in comment):
```
I'm not sure about this one. I thought we agreed not to add a new state to the system?

billsacks replied:
Good point @jgfouca . (It's possible that I was the one who suggested something like this.) So I think your point is that this should just be changed back to giving a DIFF overall status if the baselines do not exist? I'd support that. (But then again, I personally care more about the status of the individual pieces than the overall status in my parsing, so maybe someone else like @ekluzek - who I think relies more on the overall status - should weigh in on this.)

fischer-ncar replied:
I have no issue with changing it back to diff.  But I thought it was no new states to the TestStatus file.  This just changes the output from cs_status.

billsacks replied:
@fischer-ncar : @jgfouca can chime in more definitively, but I think the issue is that other tools (like test_scheduler.py) use this get_overall_test_status function, too.

billsacks commented on utils/python/CIME/test_status.py:
```diff
@@ -18,21 +18,23 @@
+MEMCOMP_PHASE          = "MEMCOMP"
+MEMLEAK_PHASE          = "MEMLEAK"
+COMPARE_PHASE          = "COMPARE" # This is one special, real phase will be COMPARE_$WHAT
+COMPARE_BASELINE_PHASE = "COMPARE_baseline"
```
I like that you distinguish this as a separate phase, addressing the request in issue #396 . However, if you make this change, I'd request:

1. Changing this code in system_tests_common.py:

```python
            self._test_status.set_status("%s_baseline" % COMPARE_PHASE, status, comments=ts_comments)
```

Now this should use COMPARE_BASELINE_PHASE in place of "%s_baseline" % COMPARE_PHASE.

2. To make this phase more distinct from the in-test comparison, I'd suggest renaming this to something like BASELINE_COMPARE - i.e., so the name looks more different from the COMPARE_whatever used for the in-test comparison.

fischer-ncar replied:
When I remove the BFAIL from the overall status, the COMPARE_BASELINE_PHASE is no longer needed.

billsacks replied:
Right, you could just revert that. However, I personally feel that having a separate COMPARE_BASELINE_PHASE is a step in the right direction. Though feel free to do whatever you prefer here - if you revert it, we can always reintroduce it later to address #396

billsacks commented on utils/python/CIME/test_status.py:
```diff
@@ -18,21 +18,23 @@
 # Special statuses that the overall test can be in
 TEST_DIFF_STATUS     = "DIFF"   # Implies a failure in one of the COMPARE phases
 NAMELIST_FAIL_STATUS = "NLFAIL" # Implies a failure in the NLCOMP phase
+TEST_BFAIL_STATUS    = "BFAIL"  # Implies baselines are missing in the COMPARE_baseline phase
```
Based on @jgfouca 's comment: It seems like there shouldn't be a separate overall status for BFAIL. However, I'd still support having "BFAIL" be defined somewhere in this file as a special string that can appear in comments, to clearly signal that this string has some special meaning.

Maybe you can introduce a new section of constants here like this:

```python
# Special strings that can appear in comments, indicating particular types of failures
TEST_NO_BASELINES_COMMENT = "BFAIL" # Implies baseline directory is missing in the baseline comparison phase
```

fischer-ncar replied:
Good idea.

fischer-ncar commented:
@jgfouca I've removed the code that had the BFAIL code for the overall status.
@billsacks I've added the new section of constants that can appear in comments.

jgfouca commented:
Waiting for @billsacks to approve and then I'll merge.
